File: algorithms/trees/avl-tree/scala/AvlTree.scala
object AvlTree {

  private case class Node(key: Int, left: Node, right: Node, height: Int)

  private def nodeHeight(node: Node): Int = if (node == null) 0 else node.height

  private def updateHeight(node: Node): Node =
    node.copy(height = 1 + math.max(nodeHeight(node.left), nodeHeight(node.right)))

  private def balanceFactor(node: Node): Int = nodeHeight(node.left) - nodeHeight(node.right)

  private def rotateRight(y: Node): Node = {
    val x = y.left
    val t2 = x.right
    val newY = updateHeight(y.copy(left = t2))
    updateHeight(x.copy(right = newY))
  }

  private def rotateLeft(x: Node): Node = {
    val y = x.right
    val t2 = y.left
    val newX = updateHeight(x.copy(right = t2))
    updateHeight(y.copy(left = newX))
  }

  private def insert(node: Node, key: Int): Node = {
    if (node == null) return Node(key, null, null, 1)
    val updated = if (key < node.key) node.copy(left = insert(node.left, key))
    else if (key > node.key) node.copy(right = insert(node.right, key))
    else return node

    val balanced = updateHeight(updated)
    val bf = balanceFactor(balanced)

    if (bf > 1 && key < balanced.left.key) return rotateRight(balanced)
    if (bf < -1 && key > balanced.right.key) return rotateLeft(balanced)
    if (bf > 1 && key > balanced.left.key)
      return rotateRight(balanced.copy(left = rotateLeft(balanced.left)))
    if (bf < -1 && key < balanced.right.key)
      return rotateLeft(balanced.copy(right = rotateRight(balanced.right)))

    balanced
  }

  private def inorder(node: Node, result: scala.collection.mutable.ListBuffer[Int]): Unit = {
    if (node == null) return
    inorder(node.left, result)
    result += node.key
    inorder(node.right, result)
  }

  def avlInsertInorder(arr: Array[Int]): Array[Int] = {
    var root: Node = null
    for (v <- arr) root = insert(root, v)
    val result = scala.collection.mutable.ListBuffer[Int]()
    inorder(root, result)
    result.toArray
  }
}

File: algorithms/trees/avl-tree/python/avl_tree.py
def avl_insert_inorder(arr: list[int]) -> list[int]:
    class Node:
        def __init__(self, key: int):
            self.key = key
            self.left: 'Node | None' = None
            self.right: 'Node | None' = None
            self.height: int = 1

    def height(node: 'Node | None') -> int:
        return node.height if node else 0

    def balance_factor(node: 'Node | None') -> int:
        return height(node.left) - height(node.right) if node else 0

    def update_height(node: Node) -> None:
        node.height = 1 + max(height(node.left), height(node.right))

    def rotate_right(y: Node) -> Node:
        x = y.left
        t2 = x.right
        x.right = y
        y.left = t2
        update_height(y)
        update_height(x)
        return x

    def rotate_left(x: Node) -> Node:
        y = x.right
        t2 = y.left
        y.left = x
        x.right = t2
        update_height(x)
        update_height(y)
        return y

    def insert(node: 'Node | None', key: int) -> Node:
        if not node:
            return Node(key)
        if key < node.key:
            node.left = insert(node.left, key)
        elif key > node.key:
            node.right = insert(node.right, key)
        else:
            return node

        update_height(node)
        bf = balance_factor(node)

        if bf > 1 and key < node.left.key:
            return rotate_right(node)
        if bf < -1 and key > node.right.key:
            return rotate_left(node)
        if bf > 1 and key > node.left.key:
            node.left = rotate_left(node.left)
            return rotate_right(node)
        if bf < -1 and key < node.right.key:
            node.right = rotate_right(node.right)
            return rotate_left(node)

        return node

    def inorder(node: 'Node | None', result: list[int]) -> None:
        if node:
            inorder(node.left, result)
            result.append(node.key)
            inorder(node.right, result)

    root = None
    for val in arr:
        root = insert(root, val)

    result: list[int] = []
    inorder(root, result)
    return result
